File: PathFindingFun Shared/Direction.swift
//
//  Direction.swift
//  PathFindingFun iOS
//
//  Created by Carl Wieland on 8/26/19.
//  Copyright © 2019 Datum Apps. All rights reserved.
//

import Foundation

enum Direction: UInt8 {
    case west = 0
    case northwest
    case northeast
    case east
    case southeast
    case southwest

    static let allDirections: [Direction] = [.west, .northwest, .northeast, .east, .southeast, .southwest]

    var isEastOrSouth: Bool {
        return self == .east || self == .southeast || self == .southwest
    }

    var isWestOrNorth: Bool {
        return self == .west || self == .northeast || self == .northwest
    }

    var reversed: Direction {
        switch self {
        case .east:
            return .west
        case .west:
            return .east
        case .southwest:
            return .northeast
        case .southeast:
            return .northwest
        case .northwest:
            return .southeast
        case .northeast:
            return .southwest
        }
    }
}

File: Unity/PathFindingFun/Assets/Scripts/Game Types/Direction.cs
using System;
using System.Runtime.CompilerServices;


public enum Direction : byte
{
    west,
    northwest,
    northeast,
    east,
    southeast,
    southwest
}

sealed class Directions
{
    private Directions()
    {
    }

    public static readonly Direction[] allDirections = {
    Direction.west,
    Direction.northwest,
    Direction.northeast,
    Direction.east,
    Direction.southeast,
    Direction.southwest
    };

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static Direction reversed(Direction dir)
    {
        switch (dir)
        {
            case Direction.east:
                return Direction.west;
            case Direction.west:
                return Direction.east;
            case Direction.southwest:
                return Direction.northeast;
            case Direction.southeast:
                return Direction.northwest;
            case Direction.northwest:
                return Direction.southeast;
            case Direction.northeast:
                return Direction.southwest;
        }
        return Direction.east;
    }

    static readonly Direction[] westNorth = { Direction.west, Direction.northwest, Direction.northeast };
    static readonly Direction[] eastSouth = { Direction.east, Direction.southeast, Direction.southwest };

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool isEastOrSouth(Direction dir)
    {
        return (int)dir > 2;
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public static bool isWestOrNorth(Direction dir)
    {
        return (int)dir < 3;

    }



}

